sys/compat/linux/linux_misc.c

```diff
@@ -2963,3 +2964,124 @@ linux_ioprio_set(struct thread *td, struct linux_ioprio_set_args *args)
 	}
 	return (error);
 }
+
+/* The only flag is O_NONBLOCK */
+#define B2L_MQ_FLAGS(bflags)	((bflags) != 0 ? LINUX_O_NONBLOCK : 0)
+#define L2B_MQ_FLAGS(lflags)	((lflags) != 0 ? O_NONBLOCK : 0)
+
+int
+linux_mq_open(struct thread *td, struct linux_mq_open_args *args)
+{
+	struct mq_attr attr;
+	int error, flags;
+
+	flags = linux_common_openflags(args->oflag);
+	if ((flags & O_ACCMODE) == O_ACCMODE || (flags & O_EXEC) != 0)
+		return (EINVAL);
+	flags = FFLAGS(flags);
+	if ((flags & O_CREAT) != 0 && args->attr != NULL) {
+		error = copyin(args->attr, &attr, sizeof(attr));
+		if (error != 0)
+			return (error);
+		attr.mq_flags = L2B_MQ_FLAGS(attr.mq_flags);
+	}
+
+	return (kern_kmq_open(td, args->name, flags, args->mode,
+	    args->attr != NULL ? &attr : NULL));
```

Comment:
What if the O_CREAT flag is not set, but args->attr is not NULL?  Wouldn't we pass a pointer to the stack garbage to kern_kmq_open?

Reply:
> What if the O_CREAT flag is not set, but args->attr is not NULL? Wouldn't we pass a pointer to the stack garbage to kern_kmq_open?

sys_kmq_open also does this.  This function is based on sys_kmq_open.  Fortunately, kern_kmq_open does the right thing.